### PR TITLE
Expand Dispatcher Wait Loop Exit Assert

### PR DESCRIPTION
**Pull request by RSalman**
For background see https://github.com/eclipse/openj9/issues/10491#issuecomment-686057277

- Existing assert expanded in `MM_ParallelDispatcher::workerEntryPoint` to cover task dispatch during shutdown
- Asserting more strictly with new assert in added else clause

Fixes: eclipse/openj9#10491

Signed-off-by: Salman Rana <salman.rana@ibm.com>

### CODE REVIEW

**Comment by RSalman** [comment]
Please review @amicic @dmitripivkine

File: gc/base/ParallelDispatcher.cpp [hunk 1]
```diff
@@ -168,10 +168,14 @@ MM_ParallelDispatcher::workerEntryPoint(MM_EnvironmentBase *env)
...
 		}
 
-		/* Thread should never exit waiting loop without being reserved with _workerThreadsReservedForGC set.
-		 * (i.e we should never have non-waiting status upon wake up with _slaveThreadsReservedForGC) */
+		/* Assert dispatcher and thread states. In most cases, a thread should exit during task dispatch (_workerThreadsReservedForGC) with a reserved status. */
 		if (_workerThreadsReservedForGC) {
-			Assert_MM_true(worker_status_reserved == _statusTable[workerID]);
+			/* Thread must exit being reserved, except in the rare case that a task is dispatched during shutdown.
+			 * In this case, _workerThreadsReservedForGC is set but the task is run single threaded and exiting threads must be dying instead of reserved. */
+			Assert_MM_true((worker_status_reserved == _statusTable[workerID]) || (!_threadsToReserve && worker_status_dying == _statusTable[workerID]));
```
Comment (dmitripivkine) [inline gc/base/ParallelDispatcher.cpp:175]: please add more `()` for both assertions

Comment (RSalman) [reply gc/base/ParallelDispatcher.cpp]: Added, although I didn't add them for single variables like `!_threadsToReserve` and `_inShutdown` in the conditions. Please let me know if they should be there and I can add them there as well.

**Comment by RSalman** [comment]
omrsockTest failing on macOS, hard to believe this change caused it.

```2020-09-08T20:50:42.5566210Z 31: /Users/runner/work/1/s/fvtest/porttest/omrsockTest.cpp:1307: Failure
2020-09-08T20:50:42.5567940Z 31:       Expected: rc
2020-09-08T20:50:42.5569130Z 31:       Which is: 1
2020-09-08T20:50:42.5569810Z 31: To be equal to: 2
2020-09-08T20:50:42.5570590Z 31: [  FAILED  ] PortSockTest.poll_functionality_many_sockets (1 ms)
2020-09-08T20:50:42.5572170Z 31: [----------] 19 tests from PortSockTest (1019 ms total)
2020-09-08T20:50:42.5572980Z 31: 
2020-09-08T20:50:42.5573700Z 31: [==========] 235 tests from 20 test cases ran. (85822 ms total)
2020-09-08T20:50:42.5574470Z 31: [  PASSED  ] 234 tests.
2020-09-08T20:50:42.5575190Z 31: [  FAILED  ] 1 test, listed below:
2020-09-08T20:50:42.5575980Z 31: [  FAILED  ] PortSockTest.poll_functionality_many_sockets
2020-09-08T20:50:42.5576680Z 31: 
2020-09-08T20:50:42.5577300Z 31:  1 FAILED TEST
2020-09-08T20:50:42.5578000Z 31:   YOU HAVE 3 DISABLED TESTS
2020-09-08T20:50:42.5578640Z 31: 
2020-09-08T20:50:42.5655400Z 31: Some textSome text
2020-09-08T20:50:42.5658720Z 36/45 Test #31: porttest ............................***Failed   85.84 sec
```

**Comment by RSalman** [comment]
Everything is passing after the restart, look like an intermittent failure unrelated to this PR

File: gc/base/ParallelDispatcher.cpp [hunk 2]
```diff
@@ -168,10 +168,14 @@ MM_ParallelDispatcher::workerEntryPoint(MM_EnvironmentBase *env)
...
 		}
 
-		/* Thread should never exit waiting loop without being reserved with _workerThreadsReservedForGC set.
-		 * (i.e we should never have non-waiting status upon wake up with _slaveThreadsReservedForGC) */
+		/* Assert dispatcher and thread states. In most cases, a thread exits during task dispatch (_workerThreadsReservedForGC) with a reserved status. */
 		if (_workerThreadsReservedForGC) {
-			Assert_MM_true(worker_status_reserved == _statusTable[workerID]);
+			/* Thread can only break out while being reserved, except in the rare case that a task is dispatched during shutdown.
+			 * In which case, _workerThreadsReservedForGC is set but the task runs single threaded and exiting threads are dying instead of reserved. */
+			Assert_MM_true((worker_status_reserved == _statusTable[workerID]) || (!_threadsToReserve && (worker_status_dying == _statusTable[workerID])));
```
Comment (amicic) [inline gc/base/ParallelDispatcher.cpp:175]: !_threadsToReserve --> (0 ==_threadsToReserve)

**Review by amicic — APPROVED** [review]

**Review by fjeremic — APPROVED** [review]

**Comment by fjeremic** [comment]
@genie-omr build all

**Pull request merged**